PR: first beta version of game challenge

koakh opened the pull request:
hello I have develop the first beta of challenge
please review it asap

thanks

buraktt commented on src/utils/main.js:
```diff
@@ -0,0 +1,53 @@
+  return `rgb(${r}, ${g}, ${b}`;
+};
+
+export const arrayIsInArray = (arr, subArr) => {
```
IMO, instead of nested loops, a combination of .some() & .every() can make this function more readable (& shorter).

koakh replied:
what a awesome recommendation
fixed with .some()

buraktt commented on src/components/grid/Grid.jsx:
```diff
@@ -0,0 +1,70 @@
+    <div className='grid'>
+      <div className='columnLabel'>
+        {columns > 0 &&
+          [...Array(columns + 1).keys()].map((e, i) => (
```
`Array.from` would be more suitable here (better readability, performance & directly iterable without the need for a spread)
(same in `Row.jsx`)

koakh replied:
another good recommendation
fixed

buraktt commented on .eslintrc.cjs:
```diff
@@ -0,0 +1,51 @@
+  },
+  extends: [
+    'plugin:react/recommended',
+    'standard',
```
this does not seem to be working. js code is not formatted in `standard` style.

koakh replied:
fixed

koakh replied:
@buraktt 
can you check if current changes fix this eslint configuration problem
thanks

buraktt commented on src/app/game.js:
```diff
@@ -0,0 +1,230 @@
+  const r = tr[0] + 1;
+
+  for (let col = t; col < b; col++) {
+    for (let row = l; row < r; row++) {
```
I agree. variable names are not really descriptive.

koakh replied:
great, 
It's rare the case I use this type of vars, thanks for pointing it out
time to make it more and more rare in future

koakh commented:
> hello @arsenii-cirplus 

- **Something wrong with the direction dice or with resolving of direction. 
I saw the situation with two points on the direction dice and it was resolved as south direction at the same time, 
what cannot be possible in any case.**

fixed: a dumb error, using wrong dice to calculate direction, thanks for pointing it out
suspect that was a refactor after some days of implementing the algorithm of direction

Also some minor bugs:

- Column and row labels look odd after resizing the window if the game is running
right I spotted it right now, 
fixed

- It's available to start a game with zero-rows board
fixed: and improve with a min rows/columns on config, and some css style

- Code-style is not JavaScript Standard Style
can you clarify, and elaborate a bit more
maybe send me a snippet with a code block before and after some changes, 
that way I can improve the code style

- There is no legend board
fixed

thanks for your great and to the point review

Mário

koakh commented on Dockerfile:
```diff
@@ -0,0 +1,14 @@
+
+ADD . .
+
+RUN npm install
```
fixed, thanks @arsenii-cirplus

koakh commented on src/app/game.js:
```diff
@@ -0,0 +1,230 @@
+  const debug = false;
+
+  const outavailableDirections = availableDirections =>
+    log(
```
I don't understand this, is related with var debug?

koakh commented on src/app/game.js:
```diff
@@ -0,0 +1,230 @@
+      let finalCol = col;
+      // work with overflow negative
+      if (row < 0) {
+        finalRow = columns - -row;
```
fixed, always good when we improve on it
thanks @arsenii-cirplus

koakh commented on src/app/game.js:
```diff
@@ -0,0 +1,230 @@
+        finalRow = columns - -row;
+      }
+      if (col < 0) {
+        finalCol = rows - -col;
```
fixed

koakh commented on src/app/game.js:
```diff
@@ -0,0 +1,230 @@
+  rows
+) => {
+  let result = position;
+  switch (directionStatus[0]) {
```
I don't get it, can you elaborate maybe with a snippet
thanks @arsenii-cirplus

koakh commented on src/utils/main.js:
```diff
@@ -0,0 +1,53 @@
+  min = Math.ceil(min);
+  max = Math.floor(max);
+  min = Math.ceil(min);
+  max = Math.floor(max);
```
fixed 
sure it was a typo, a copy paste one
thanks

koakh commented on src/App.jsx:
```diff
@@ -0,0 +1,257 @@
+
+  // eventHandlers
+  const onClickStartGameHandler = e => {
+    setTimeout(() => {});
```
fixed, seems a lost line forget my mistake
thanks for pointing it out

koakh commented on src/App.jsx:
```diff
@@ -0,0 +1,257 @@
+  // effects
+  useEffect(() => {
+    if (state.gameStatus.running && isBootleTurn) {
+      onClickRollDiceHandler();
```
I debug the code

and all intervals are cleared after players moving, none is stalled...and don't have any side effect
(I leave a variable const debugTimeInterval = false, that if enabled we can observe it)
...this is managed inside onClickRollDiceHandler setInterval when it detects player stop's

I understand that this must be a hack way and not a hooks idiomatic way, 
but to implement this delay in player movements, like a sort of animation, I must use a setInterval and in this case conflicts with your wise recommendation and best practices (not leave stalled intervals running)

but if I clear the interval in useEffect cleanup the player stop because, we clear the interval, 
seems that the solution is not use timers to move players without delay, and use a simple loop without any sort of delay, or thrust that clear intervals are well implemented in logic

I completely agree that this is far from optional..........works but not perfect

or try to figure out a approach that works in both cases :)

do you have any suggestion @buraktt 

what I think is a requirement, get rid of timer and move players without delay, but in my humble opinion I like the way delay is used, else players will very fast

waiting for your advice and recommendation in this recommendation

koakh replied:
Hello Burak

I already do the clean up in the weekend, like you suggested, but if I clean up the interval on unmount, player movement stops because it is controlled by that time interval

I try to explain that in push review comment, if I clear interval, stop occurs, and game stop creating other side effects,  that's why I request some advice

For what I see in that debug, the only way to get rid of set interval is refactor all code and move the players from position x to y, with one step only, and lost this delays and animation, like player moves one step from x to x milliseconds......

I don't know how others that do the challenge did the delay and animation without unmount, and without timmers...to be honest I can't figure it out
Maybe use the timer outside of component in game.js

I can try refactor and get rid of interval in component, and start from that point to see if some idea occurs

About memory leaks, I debug everything and when players stop moving, all intervals are clear, not in unmounts but when players stop move

I understand that this is a hack and to far from ideal

Waiting from some advice here

Like refactor code, or get rid of interval in component

Thanks

koakh replied:
FIXED with great recommendation of @realiarthur

koakh commented on src/App.jsx:
```diff
@@ -1,53 +1,70 @@
+  const [state, dispatch] = useStateValue()
+  const [uiDisabled, setUIDisabled] = useState(false)
+  const [isBootleTurn, setIsBottleTurn] = useState(false)
+  // TODO: this is a temporary state until figure out a solution for Burak recommendation
```
Hello @buraktt 
I don't understand update for what?
Remove that todo comment?
Sorry to ask and don't understand
Thanks

buraktt replied:
@koakh sorry, I missed your other comment since you marked it as resolved. I'll write under that.

koakh replied:
thanks @buraktt

realiarthur commented on src/App.jsx:
```diff
@@ -0,0 +1,294 @@
+        }
+      })
+
+      // check player positions to define end of game
```
it is better to split the logic of moving, victory checking, etc. into different functions or methods. 
Also, it is better to keep it in a separate file, not to overload this component and this handler

koakh replied:
FIXED extracted two functions from onClickRollDiceHandler
rollDice() and checkEndGame()

onClickRollDiceHandler is a lot more clean

```js
  const onClickRollDiceHandler = () => {
    rollDice(state, dispatch, setUIDisabled, rollInterval)
  }
```

realiarthur commented on src/App.jsx:
```diff
@@ -0,0 +1,294 @@
+  const [uiDisabled, setUIDisabled] = useState(false)
+  const [isBootleTurn, setIsBottleTurn] = useState(false)
+  // TODO: this is a temporary state until figure out a solution for Burak recommendation
+  const [rollInterval, setRollInterval] = useState(null)
```
it is better to use `useRef` to store `rollInterval` to prevent unnecessary rerender on state changing

koakh replied:
yes sure, another awesome recommendation :)

FIXED and a lot better in render counts

realiarthur commented on src/components/grid/Grid.jsx:
```diff
@@ -0,0 +1,74 @@
+    });
+  };
+
+  const resizeHandler = useCallback(() => {
```
there is no point to use `useCallback` here. Just create this function inside the `useEffect` to use the same reference for cleanup function

koakh replied:
nice recommendation again

refactored to 
```js
  useEffect(() => {
    const resizeHandler = () => {
      // block change while running
      if (running) {
        return
      }
      renderGrid()
    }

    renderGrid()
    window.addEventListener('resize', resizeHandler)
    return () => {
      window.removeEventListener('resize', resizeHandler)
    }
  }, [running])
```

I think is what you mean, just to be sure that there is not a misinterpretation

realiarthur commented on src/components/grid/Grid.jsx:
```diff
@@ -0,0 +1,74 @@
+
+  useEffect(() => {
+    renderGrid();
+    window.addEventListener('resize', resizeHandler);
```
in production-ready solutions, it is better to use debounce for resize event callback

koakh replied:
sure, fixed with a debounce of 500ms
thanks for pointing it out @realiarthur

realiarthur commented on src/components/cell/Cell.jsx:
```diff
@@ -0,0 +1,120 @@
+  const [x, y] = position;
+  const [state] = useStateValue();
+
+  const drawThing = (x, y, player, bottle, showPos = false) => {
```
nit: it is better to render the player and bottle directly, instead of checking each cell for the presence

And the same for gpgp and protected zone

koakh replied:
this will need a big refactor in whole render cells approach
it is a requirement or a suggestion?

realiarthur replied:
just a suggestion :)

koakh replied:
great

I hope you don't mind but I'll leave it as it is, 
this change would radically change the way I implemented the feature, 
and that would force me to redo all the code and delay the challenge even more

thanks for understanding @realiarthur

realiarthur commented on src/app/game.js:
```diff
@@ -0,0 +1,222 @@
+    for (let row = lft; row < rgt; row++) {
+      let finalRow = row
+      let finalCol = col
+      // work with overflow negative
```
you use this check at least twice in the app. It is better to extract it into a function or method

koakh replied:
I don't understand this recommendation
I only see it one time at src/app/game.js line 74, like you mention

can you point me the other place is is used

thanks @realiarthur

realiarthur replied:
I thought that lines #175 - #190 do the same check. Am I wrong?

koakh commented:
Hello Burak

I already do the clean up in the weekend, like you suggested, but if I clean up the interval on unmount, player movement stops because it is controlled by that time interval

I try to explain that in push review comment, if I clear interval, stop occurs, and game stop creating other side effects,  that's why I request some advice

For what I see in that debug, the only way to get rid of set interval is refactor all code and move the players from position x to y, with one step only, and lost this delays and animation, like player moves one step from x to x milliseconds......

I don't know how others that do the challenge did the delay and animation without unmount, and without timmers...to be honest I can't figure it out
Maybe use the timer outside of component in game.js

I can try refactor and get rid of interval in component, and start from that point to see if some idea occurs

About memory leaks, I debug everything and when players stop moving, all intervals are clear, not in unmounts but when players stop move

I understand that this is a hack and to far from ideal

Waiting from some advice here

Like refactor code, or get rid of interval in component

Thanks

On Oct 31, 2022, 15:30, at 15:30, Burak Topbastekin ***@***.***> wrote:
***@***.*** commented on this pull request.
>
>
>
>> +import { getDirection, getMoveStepPositions } from './app/game';
>+import { useStateValue } from './app/state/useStateValue';
>+import Dice from './components/dice/Dice';
>+import Grid from './components/grid/Grid';
>+import { arrayIsInArray, log, randomNumber } from './utils/main';
>+
>+const App = () => {
>+  // hooks
>+  const [state, dispatch] = useStateValue();
>+  const [uiDisabled, setUIDisabled] = useState(false);
>+  const [isBootleTurn, setIsBottleTurn] = useState(false);
>+
>+  // effects
>+  useEffect(() => {
>+    if (state.gameStatus.running && isBootleTurn) {
>+      onClickRollDiceHandler();
>
***@***.*** any function set to run later (eg: an event listener,
>`setTimeout` ) inside `useEffect` should be cleaned if the component
>**unmounts**. Otherwise it will create a memory leak.
>
>The link I've added already includes the solution but let me simplify
>it.
>
>eg:
>
>```
>useEffect(() => {
>  // set interval
>  return () => {
>    // clear interval
>  } 
>}, [])
>```
>so, in this case:
>
>```
>useEffect(() => {
>  if (state.gameStatus.running && isBootleTurn) {
>    onClickRollDiceHandler(); // this call sets an interval
>  }
>  ...
>  ...
>  return () => {
>     // interval set by onClickRollDiceHandler should be cleaned here
>  } 
>}, [...])
>```
>
>
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/koakh/ChallengeTheLastBottle/pull/1#discussion_r1009557459
>You are receiving this because you were mentioned.
>
>Message ID:
***@***.***>

realiarthur commented:
[koakh](https://github.com/koakh) you can't clean up the effect properly because you have `state` as a dependency of `useEffect`. Of course, `state` changed after the first move which triggers the cleanup function. You can use `state.gameStatus.running` in the dependency array and it will let you to clean interval properly

koakh commented:
Awesome tip, I confess I don't try this :)
I try it later
Thanks @realiarthur

koakh commented on src/App.jsx:
```diff
@@ -0,0 +1,294 @@
+    }
+    return () => {
+      if (debugTimeInterval) {
+        log(`useEffect: cleanUp interval ${rollInterval}`)
```
yes is a wip
just some debug until I fix the clear interval in useEffect,
already removed all references to debugTimeInterval and its console logs
@buraktt already point that

thanks @realiarthur 

FIXED

koakh commented on src/hooks/useElementSize.js:
```diff
@@ -0,0 +1,46 @@
+)
+
+export const useElementSize = ref => {
+  const elRef = useRef(null)
```
It is used bellow in useLayoutEffect
but seems that I don't undertstand this 
can you elaborate 
```js
export const useElementSize = ref => {
  const elRef = useRef(null)
  const [{ height, width }, setSize] = useState(() => ({
    width: 0,
    height: 0
  }))

  useLayoutEffect(() => {
    const element = elRef.current || ref?.current
  ...
```
thanks

realiarthur replied:
yes, you are right! sorry

koakh commented on src/App.jsx:
```diff
@@ -0,0 +1,294 @@
+  // hooks
+  const [state, dispatch] = useStateValue()
+  const [uiDisabled, setUIDisabled] = useState(false)
+  const [isBootleTurn, setIsBottleTurn] = useState(false)
```
is just a helper, but refactored code as suggested and it is a lot better
 FIXED as suggested
